crates/proof-compression/src/chain.rs

```diff
@@ -153,64 +140,96 @@ where
     SnarkWrapperProof::FFfonk(final_proof)
 }
 
-pub fn precompute_proof_chain_with_fflonk<BS>(blob_storage: &BS)
-where
-    BS: BlobStorageExt,
-{
-    let context_manager = SimpleContextManager::new();
+pub fn precompute_proof_chain_with_fflonk(setup_data_cache: Arc<dyn CompressorBlobStorageExt>) {
     <FflonkSnarkWrapper as SnarkWrapperStep>::run_pre_initialization_tasks();
-    let compact_raw_crs =
-        <FflonkSnarkWrapper as SnarkWrapperStep>::load_compact_raw_crs(blob_storage);
 
     let start = std::time::Instant::now();
-    CompressionMode1::precomputae_and_store_compression_circuits(blob_storage, &context_manager);
-    CompressionMode2::precomputae_and_store_compression_circuits(blob_storage, &context_manager);
-    CompressionMode3::precomputae_and_store_compression_circuits(blob_storage, &context_manager);
-    CompressionMode4::precomputae_and_store_compression_circuits(blob_storage, &context_manager);
-    CompressionMode5ForWrapper::precomputae_and_store_compression_circuits(
-        blob_storage,
-        &context_manager,
+
+    let (precomputation, vk, finalization_hint) = CompressionMode1::precompute_compression_circuits(
+        setup_data_cache
+            .get_compression_mode1_previous_vk()
+            .unwrap(),
+    );
+    setup_data_cache
+        .set_compression_mode1_setup_data(&precomputation, &vk, &finalization_hint)
+        .context("Failed to set compression mode 1 setup data")
+        .unwrap();
+
+    let (precomputation, vk, finalization_hint) = CompressionMode2::precompute_compression_circuits(
+        setup_data_cache
+            .get_compression_mode2_previous_vk()
+            .unwrap(),
+    );
+    setup_data_cache
+        .set_compression_mode2_setup_data(&precomputation, &vk, &finalization_hint)
+        .context("Failed to set compression mode 2 setup data")
```

Comment:
This looks a bit strange(and in other places). Context will wrap the error into the `anyhow` error, but you are doing unwrap later. I'd suggest either to make function return `anyhow::Result`(preferably IMO, I don't like that we have such a big amount of explicit unwraps in crypto codebase), or to use `.expect`